ibis/backends/base/sql/alchemy/query_builder.py

```diff
@@ -87,6 +88,33 @@ def _format_table(self, expr):
                 ref_op.name,
                 *(sa.column(n, to_sqla_type(t)) for n, t in schema.items()),
             )
+        elif isinstance(ref_op, ops.View):
+            from sqlalchemy_views import CreateView, DropView
+
+            compiled = ref_op.child.compile()
+
+            if isinstance(ref_op, ops.SQLStringView):
+                definition = sa.text(ref_op.query)
+            else:
+                definition = compiled.compile()
+
+            result = view = sa.table(
+                ref_op.name,
+                *(
+                    sa.column(n, to_sqla_type(t))
+                    for n, t in ref_expr.schema().items()
+                ),
+            )
+
+            create_view = CreateView(view, definition, or_replace=True)
+            with ref_op.backend.con.connect() as con:
+                con.execute(create_view)
+
+            # drop the view just before interpreter shutdown
+            atexit.register(
```

Comment:
:wave: Apologies for the drive-by comment here. I saw a link to this on twitter and was curious about the implementation.

One thing to maybe be nervous about w/r/t cleaning up using `atexit` is that Python can't guarantee that `atexit` handlers actually run. In particular, they won't run in the event of a power failure or if the process receives a `SIGKILL`.

I'm not sure about the various other backends that Ibis supports, but postgres at least has the concept of a temporary view, which is automatically destroyed when the postgres session ends. Using that might be safer for postgres to ensure that created views don't outlive the process.

Reply:
@ssanderson Yeah, I actually used `CREATE TEMPORARY VIEW` first, but didn't generalize it so that a backend could customize temporary view creation.

I've now made it so that each backend has to implement a `_create_temporary_view` method, which turned out nice since I can remove the `sqlalchemy-views` dependency since only postgres and duckdb from the sqlalchemy backends support it right now.

Reply:
In the future though, I don't know if there's much better we can do than `atexit` for databases that don't have temporary view creation.

I suppose we could clean up using the `after_execute` method, but then there's still the case of `.sql` and `.alias` expressions that were created but never executed to deal with.

I don't know if there's a robust way to solve this without the help of the database.

Reply:
> I don't know if there's a robust way to solve this without the help of the database.

Makes sense. You're fundamentally creating an entity that lives outside of the process, so if you want to be robust to sudden death of the process, the thing that handles that death needs to live outside the process.

> I don't know if there's a robust way to solve this without the help of the database.

The other possible option seems like it would be to wrap the custom sql in a CTE and attach it to the query in a way that makes the scoping work out with the desired Ibis semantics? The nice thing about that would be that you wouldn't have to deal with the implications of a semantically readonly operation having side-effects in the database (I could imagine this causing other headaches w/r/t permissions, or name collisions between multiple users). Without knowing more about Ibis' architecture, I'm not sure how hard or easy that would be though.

Reply:
> Makes sense. You're fundamentally creating an entity that lives outside of the process, so if you want to be robust to sudden death of the process, the thing that handles that death needs to live outside the process.

Indeed.

> The other possible option seems like it would be to wrap the custom sql in a CTE and attach it to the query in a way that makes the scoping work out with the desired Ibis semantics? The nice thing about that would be that you wouldn't have to deal with the implications of a semantically readonly operation having side-effects in the database (I could imagine this causing other headaches w/r/t permissions, or name collisions between multiple users). Without knowing more about Ibis' architecture, I'm not sure how hard or easy that would be though.

I went down this path initially 😄, and I agree it would be the ideal solution. It was pretty challenging and I couldn't come up with something that I was happy with.

Reply:
Alright, I was able to cut out one temporary view creation out by using CTEs for the `.sql` method. We still need the the temp view for `.alias` for now, since any `.sql` expressions that use the result of `.alias` won't have the entire query available (it does, but the architecture doesn't support glomming these together without a pile of hacks).

With that, I think I am done being nerd sniped for now 😂.